Improve tracks menu UI speed

rt1shnik opened the pull request:
Closes https://github.com/PDApps/DICTI/issues/176

Время от клика до появления интерфейса в ландшафтной ориентации (3 варианта аудио, 2 варианта субтитров) / время от onCreate до появления UI. Was:
```
Samsung Note 4 (debug) ~195 ms / ~170 ms
Xiaomi 11T, Android 12 (debug) ~258 ms / ~226 ms
```
Now it:
```
Samsung Note 4 (debug) ~143 ms / ~121 ms
Xiaomi 11T, Android 12 (debug) ~237 ms / ~178 ms
```

<details>
  <summary>Note 4 profiler</summary>

[Note_4_tracks_land_commit_1_cpu-art-20220525T160612.zip](https://github.com/PDApps/vlc-android/files/8771378/Note_4_tracks_land_commit_1_cpu-art-20220525T160612.zip)
![Screenshot from 2022-05-25 16-06-04](https://user-images.githubusercontent.com/4016063/170269091-e03f3593-50b2-4420-afc0-11a437fcbafb.png)

</details>

Во время профайлинга на t11 я заметил, что UI появляется сначала на мгновение в раскрытом виде, а потом в свернутом варианте. По профайлеру тоже видно, что после первого расчета UI запускается ещё один прощет.  
<details>
  <summary>T11 profiler</summary>

[t11_tracks_land_commit_1_cpu-art-20220525T161626.zip](https://github.com/PDApps/vlc-android/files/8771458/t11_tracks_land_commit_1_cpu-art-20220525T161626.zip)
![Screenshot from 2022-05-25 16-16-14](https://user-images.githubusercontent.com/4016063/170271259-0028804a-8041-4639-8e3d-1fc5e3ae60bd.png)

</details>

Исправил мелькание https://github.com/PDApps/vlc-android/pull/1/commits/8af12596c7995551b528364c2195a748dfc39673 (панели иногда появлялись сначала в развернутом состоянии).
Заменил ещё один `ConstraintLayout` 209406a0883c02b41c6b4f5d47373b9aa36cbd6e.
```
Samsung Note 4 (debug) ~129 ms / ~107 ms
Samsung Note 4 (release) ~208 ms / ~177 ms
Xiaomi 11T, Android 12 (debug) ~220 ms / ~175 ms (55-330 ms)
Xiaomi 11T, Android 12 (release) ~206 ms (150-300 ms) / ~158 ms (60-260 ms)
```
<details>
  <summary>Note 4 profiler</summary>

[Note_4_tracks_land_commit_3_cpu-art-20220526T093148.zip](https://github.com/PDApps/DICTI/files/8776818/Note_4_tracks_land_commit_3_cpu-art-20220526T093148.zip)
![Screenshot from 2022-05-26 09-25-48](https://user-images.githubusercontent.com/4016063/170431591-a4c4fca6-aa4b-407e-9957-30d9b9316719.png)

</details>

<details>
  <summary>T11 profiler</summary>

[t11_tracks_land_commit_3_cpu-art-20220526T094634.zip](https://github.com/PDApps/vlc-android/files/8776886/t11_tracks_land_commit_3_cpu-art-20220526T094634.zip)
![Screenshot from 2022-05-26 09-46-20](https://user-images.githubusercontent.com/4016063/170433337-90dc2a05-f64f-4dbb-80a7-ab3e013fb8ba.png)

</details>

По профайлеру видно, что в Note 4 появилось ещё одна секция `measure`. Думаю, это из-за моего исправления 8af12596c7995551b528364c2195a748dfc39673. Я там специально вызываю `measure`, чтобы узнать высоту панелей. До этого приложение дожидалось пока просчитаются размеры панели в раскрытом состоянии, и только потом устанавливало нулевую высоту. На t11 наоборот пропал один кусок `measure` в конце.
Я временно удалил свой `measure`, но вторая секция `measure` осталась в Note 4. Это не мой `measure` был. Что-то странное, буду в первую очередь ориентироваться на замеры производительности. Ещё я обнаружил, что старая и новая версия профайлера показывают немного разные данные.

Сдалал, чтобы опции создавались при первом их показе, а не при создании главного диалога.
```
Samsung Note 4 (debug) ~112 ms / ~92 ms
```
<details>
  <summary>Note 4 profiler</summary>

[Note_4_tracks_land_commit_4_cpu-art-20220526T115212.zip](https://github.com/PDApps/vlc-android/files/8777855/Note_4_tracks_land_commit_4_cpu-art-20220526T115212.zip)
![Screenshot from 2022-05-26 11-52-02](https://user-images.githubusercontent.com/4016063/170458276-9d644cdf-aede-4e4f-8e77-bb4ab5ad7350.png)

</details>

<details>
  <summary>T11 profiler</summary>

[t11_tracks_land_commit_4_cpu-art-20220526T120053.zip](https://github.com/PDApps/vlc-android/files/8777853/t11_tracks_land_commit_4_cpu-art-20220526T120053.zip)
![Screenshot from 2022-05-26 12-00-28](https://user-images.githubusercontent.com/4016063/170458294-0b4e154b-1495-4f2e-81e0-33b169f0d6f0.png)

</details>

Вынес секцию с вариантами видео в `ViewStub`, так как в 99% случаев она все равно будет спрятана.
```
Samsung Note 4 (debug) ~99 ms / ~81 ms
Samsung Note 4 (release) ~142 ms / ~120 ms
Xiaomi 11T, Android 12 (release) ~199 ms (95-330 ms) / не считал вторую цифру, так как пока считаю первую все логи почему-то затираются. Да и разброс значений на T11 огромный (минимальное значение отличается от максимального в несколько раз), что не способствует пониманию, улучшилось что-то или нет. На Note 4 разброс значений +-20%.
Xiaomi 11T, Android 12 (release) ещё раз протестировал ~239 ms (160-330 ms)
```

<details>
  <summary>Note 4 profiler</summary>

[Note_4_tracks_land_commit_5_cpu-art-20220526T145604.zip](https://github.com/PDApps/vlc-android/files/8778732/Note_4_tracks_land_commit_5_cpu-art-20220526T145604.zip)
![Screenshot from 2022-05-26 14-55-54](https://user-images.githubusercontent.com/4016063/170483565-294cc624-1344-4bef-8a9f-0dbf526011ff.png)
</details>

rt1shnik committed:
Measure VideoTracksDialog performance

rt1shnik committed:
Use LinearLayout for video_track_item.xml

This modification boost UI speed. 
"Время от клика до появления интерфейса в ландшафтной ориентации (3 варианта аудио, 2 варианта субтитров) / время от onCreate до появления UI." Was:
Samsung Note 4 (debug) ~195 ms / ~170 ms
Xiaomi 11T, Android 12 (debug) ~258 ms / ~226 ms
Now it:
Samsung Note 4 (debug) ~143 ms / ~121 ms
Xiaomi 11T, Android 12 (debug) ~237 ms / ~178 ms

rt1shnik committed:
Avoid UI jumping when open tracks

"Во время профайлинга на t11 я заметил, что UI появляется сначала на мгновение в раскрытом виде, а потом в свернутом варианте. По профайлеру тоже видно, что после первого расчета UI запускается ещё один прощет."

voronnate commented:
> `Samsung Note 4 (debug) ~195 ms / ~170 ms Xiaomi 11T, Android 12 (debug) ~258 ms / ~226 ms`

На более новом устройстве время на треть дольше? А вы говорили, что в Note 4 процессор более слабый.

Как я понял, "ничего подозрительного" (очевидного) нет? Нужно только изучать профилировщик?

> Во время профайлинга на t11 я заметил, что UI появляется сначала на мгновение в раскрытом виде, а потом в свернутом варианте. По профайлеру тоже видно, что после первого расчета UI запускается ещё один прощет. 

Похоже, VLC для Android писали не особо хорошие программисты. То каждую секунду пересчет интерфейса, то двойное открытие диалога... Так нам придется много чего дописывать.

rt1shnik commented:
@voronnate 
> На более новом устройстве время на треть дольше? А вы говорили, что в Note 4 процессор более слабый.

Да, в этом случае T11 медленнее. Но тут много факторов может влиять. Версия Android, модификации прошивки от производителей телефонов, ABI (T11 поддерживает arm64-v8a). Плюс, я тестировал debug сборки. Скорее всего, с release сборкой результат будет отличаться, я это ещё проверю.

> Как я понял, "ничего подозрительного" (очевидного) нет? Нужно только изучать профилировщик?

В этом случае да.

> Похоже, VLC для Android писали не особо хорошие программисты. То каждую секунду пересчет интерфейса, то двойное открытие диалога... Так нам придется много чего дописывать.

Над Android версией не так уж много людей работает. Большинство изменений делает один человек. Несколько лет назад это ещё был Geoffrey Métais, потом на смену ему пришел Nicolas Pomepuy. Все проблемы, над которыми я работал, появились после его изменений. Он не сильно следит за производительностью.

> Так нам придется много чего дописывать.

Это лучше, чем с нуля все писать. Не так уж много есть нормальных проигрывателей с открытым исходным кодом.

rt1shnik committed:
Use LinearLayout for player_overlay_track_item.xml

This modification boost UI speed. 
"Время от клика до появления интерфейса в ландшафтной ориентации (3 варианта аудио, 2 варианта субтитров) / время от onCreate до появления UI." Was:
Samsung Note 4 (debug) ~143 ms / ~121 ms
Xiaomi 11T, Android 12 (debug) ~237 ms / ~178 ms
Now it:
Samsung Note 4 (debug) ~129 ms / ~107 ms
Xiaomi 11T, Android 12 (debug) ~220 ms / ~175 ms (55-330 ms)

rt1shnik committed:
Remove redundant import

rt1shnik committed:
Move options creating to the moment when user open them

This modification boost UI speed. 
Время от клика до появления интерфейса в ландшафтной ориентации (3 варианта аудио, 2 варианта субтитров) / время от onCreate до появления UI. Was:
Samsung Note 4 (debug) ~129 ms / ~107 ms
Xiaomi 11T, Android 12 (debug) ~220 ms / ~175 ms (55-330 ms)(60-260 ms)
Now it:
Samsung Note 4 (debug) ~112 ms / ~92 ms
Xiaomi 11T, Android 12 (debug) не успел замерить, разрядилась батарея

rt1shnik committed:
Use LinearLayout for player_overlay_tracks.xml

Minor UI speed boost.
Время от клика до появления интерфейса в ландшафтной ориентации (3 варианта аудио, 2 варианта субтитров) / время от onCreate до появления UI. Was:
Samsung Note 4 (debug) ~112 ms / ~92 ms
Now it:
Samsung Note 4 (debug) ~110 ms / ~89 ms

rt1shnik committed:
Use ViewStub for video tracks segment

This modification boost UI speed. 
Время от клика до появления интерфейса в ландшафтной ориентации (3 варианта аудио, 2 варианта субтитров) / время от onCreate до появления UI. Was:
Samsung Note 4 (debug) ~112 ms / ~92 ms
Now it:
Samsung Note 4 (debug) ~99 ms / ~81 ms
Xiaomi 11T, Android 12 (release) ~199 ms (95-330 ms)
